src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs

```diff
@@ -1722,11 +1729,51 @@ var outboundNavigations
 
             foreach (var navigation in outboundNavigations)
             {
-                var addedIncludeTreeNode = includeTreeNode.AddNavigation(navigation);
-                PopulateEagerLoadedNavigations(addedIncludeTreeNode);
+                includeTreeNode.AddNavigation(navigation);
             }
         }
 
+        private void VerifyNoAutoIncludeCycles(
+            IEntityType entityType, HashSet<IEntityType> visitedEntityTypes, INavigationBase previousNavigation)
+        {
+            if (_nonCyclicAutoIncludeEntityTypes.Contains(entityType))
+            {
+                return;
+            }
+
+            if (!visitedEntityTypes.Add(entityType))
+            {
+                throw new InvalidOperationException(CoreStrings.AutoIncludeNavigationCycle(
+                    entityType.ShortName(),
+                    visitedEntityTypes.Select(e => e.ShortName()).Join()));
+            }
+
+            var autoIncludedNavigations = GetOutgoingEagerLoadedNavigations(entityType)
+                .Where(n => !(n is INavigation navigation && navigation.ForeignKey.IsOwnership));
+
+            foreach (var navigationBase in autoIncludedNavigations)
+            {
+                if (previousNavigation?.Inverse == navigationBase
+                    && navigationBase is INavigation)
+                {
+                    continue;
+                }
+
+                VerifyNoAutoIncludeCycles(navigationBase.TargetEntityType, visitedEntityTypes, navigationBase);
+            }
+
+            _nonCyclicAutoIncludeEntityTypes.Add(entityType);
```

Comment:
```suggestion
            _nonCyclicAutoIncludeEntityTypes.Add(entityType, previousNavigation);
```

Change `_nonCyclicAutoIncludeEntityTypes` do dictionary and now you can reconstruct the cycle (more or less)